core/application/bloodDonationWorkflow/DonorSearchService.ts

```diff
@@ -0,0 +1,180 @@
+import {
+  DonationStatus,
+  DonorSearchDTO
+} from '../../../commons/dto/DonationDTO'
+import Repository from '../models/policies/repositories/Repository'
+import { getGeohashNthNeighbors } from '../utils/geohash'
+import {
+  QueryConditionOperator,
+  QueryInput
+} from '../models/policies/repositories/QueryTypes'
+import {
+  DonorRoutingAttributes,
+  StepFunctionInput
+} from './Types'
+import { StepFunctionModel } from '../models/stepFunctions/StepFunctionModel'
+import { DONOR_SEARCH_PK_PREFIX } from '../models/dbModels/DonorSearchModel'
+import { AcceptedDonationFields } from '../models/dbModels/AcceptDonationModel'
+import { LocationDTO, UserDetailsDTO } from '../../../commons/dto/UserDTO'
+import { getBloodRequestMessage } from './BloodDonationMessages'
+import LocationModel from '../models/dbModels/LocationModel'
+
+export class DonorSearchService {
+  async routeDonorRequest(
+    donorRoutingAttributes: DonorRoutingAttributes,
+    sourceQueueArn: string,
+    userProfile: UserDetailsDTO,
+    donorSearchRepository: Repository<DonorSearchDTO>,
+    stepFunctionModel: StepFunctionModel
+  ): Promise<void> {
+    const { seekerId, requestPostId, createdAt } = donorRoutingAttributes
+
+    const donorSearchRecord = await donorSearchRepository.getItem(
+      `${DONOR_SEARCH_PK_PREFIX}#${seekerId}`,
+      `${DONOR_SEARCH_PK_PREFIX}#${createdAt}#${requestPostId}`
+    )
+
+    if (donorSearchRecord === null) {
+      await donorSearchRepository.create({
+        id: donorRoutingAttributes.requestPostId,
+        ...donorRoutingAttributes,
+        status: DonationStatus.PENDING,
+        retryCount: 0
+      })
+    } else if (donorSearchRecord.status === DonationStatus.COMPLETED) {
+      if (
+        sourceQueueArn === process.env.DONOR_SEARCH_QUEUE_ARN &&
+        donorRoutingAttributes.bloodQuantity > donorSearchRecord.bloodQuantity
+      ) {
+        const updatedRecord: Partial<DonorSearchDTO> = {
+          ...donorRoutingAttributes,
+          id: requestPostId,
+          status: DonationStatus.PENDING,
+          retryCount: 0
+        }
+        await donorSearchRepository.update(updatedRecord)
+      }
+    }
+
+    const retryCount = donorSearchRecord?.retryCount ?? 0
+    const updatedRecord: Partial<DonorSearchDTO> = {
+      ...donorRoutingAttributes,
+      id: requestPostId,
+      retryCount: retryCount + 1
+    }
+
+    if (retryCount >= Number(process.env.MAX_RETRY_COUNT)) {
+      updatedRecord.status = DonationStatus.COMPLETED
+      await donorSearchRepository.update(updatedRecord)
```

Comment:
you could just 
```ts
return donorSearchRepository.update(updatedRecord)
```
assuming `donorSearchRepository.update(updatedRecord)` also returns void

Reply:
`donorSearchRepository.update(updatedRecord)` returns a DTO

Reply:
owh. lets leave it as is in that case boss